app/controllers/fhir/r4/api_controller.rb

```diff
@@ -0,0 +1,374 @@
+    doorkeeper_authorize! :read
+    check_api_access
+  end
+  before_action :cors_headers
+
+  # Return a resource given a type and an id.
+  #
+  # Supports (reading): Patient, Observation, QuestionnaireResponse
+  #
+  # GET /[:resource_type]/[:id]
+  def show
+    resource_type = params.permit(:resource_type)[:resource_type]&.downcase
+    case resource_type
+    when 'patient'
+      resource = get_patient(params.permit(:id)[:id])
+    when 'observation'
+      resource = get_laboratory(params.permit(:id)[:id])
+    when 'questionnaireresponse'
+      resource = get_assessment(params.permit(:id)[:id])
+    else
+      status_bad_request && return
+    end
+
+    status_forbidden && return if resource.nil?
+
+    status_ok(resource.as_fhir) && return
+  rescue StandardError
+    render json: operation_outcome_fatal.to_json, status: :internal_server_error
+  end
+
+  # Update a resource given a type and an id.
+  #
+  # Supports (updating): Patient
+  #
+  # PUT /fhir/r4/[:resource_type]/[:id]
+  def update
+    # Parse in the FHIR::Patient
+    contents = FHIR.from_contents(request.body.string)
+    status_bad_request && return if contents.nil? || !contents.valid?
+
+    resource_type = params.permit(:resource_type)[:resource_type]&.downcase
+    case resource_type
+    when 'patient'
+      updates = Patient.from_fhir(contents).select { |_k, v| v.present? }
+      resource = get_patient(params.permit(:id)[:id])
+    else
+      status_bad_request && return
+    end
+
+    status_forbidden && return if resource.nil?
+
+    # Try to update the resource
+    status_bad_request && return if updates.nil? || !resource.update(updates)
+
+    status_ok(resource.as_fhir) && return
+  rescue StandardError
+    render json: operation_outcome_fatal.to_json, status: :internal_server_error
+  end
+
+  # Create a resource given a type.
+  #
+  # Supports (writing): Patient
+  #
+  # POST /fhir/r4/[:resource_type]
+  def create
+    # Parse in the FHIR::Patient
+    contents = FHIR.from_contents(request.body.string)
+    status_bad_request && return if contents.nil? || !contents.valid?
+
+    resource_type = params.permit(:resource_type)[:resource_type]&.downcase
+    case resource_type
+    when 'patient'
+      # Construct a Sara Alert Patient
+      resource = Patient.new(Patient.from_fhir(contents))
+
+      # Responder is self
+      resource.responder = resource
+
+      # Creator is authenticated user
+      resource.creator = current_resource_owner
+
+      # Jurisdiction is the authenticated user's jurisdiction
+      resource.jurisdiction = current_resource_owner.jurisdiction
+
+      # Generate a submission token for the new monitoree
+      resource.submission_token = SecureRandom.hex(20) # 160 bits
+    end
+
+    status_bad_request && return if resource.nil?
+
+    status_bad_request && return unless resource.save
+
+    if resource_type == 'patient'
+      # Send enrollment notification
+      resource.send_enrollment_notification
```

Comment:
send_enrollment_notification relies on the preferred_contact_method, how does preferred_contact_method get a value by the time we get here?

Reply:
I think this should come from telecom, or have it be a separate extension. Will look into it.